BedrockServer.cpp

```diff
@@ -288,41 +288,61 @@ void BedrockServer::sync(SData& args,
+
+                        // This command completed in peek, respond to it appropriately, either directly or by sending it
+                        // back to the sync thread.
+                        SASSERT(command.complete);
+                        if (command.initiatingPeerID) {
+                            command.finalizeTimingInfo();
+                            syncNode.sendResponse(command);
+                        } else {
+                            server._reply(command);
+                        }
+                        continue;
                     }
-                    continue;
+                } else {
+                    // If the command had an https request before we tried to peek it, we record that, as we'll need to
+                    // restart the transaction.
+                    alreadyHadHttpsRequest = true;
                 }
-            }
 
-            // If we've dequeued a command with an incomplete HTTPS request, we move it to httpsCommands so that every
-            // subsequent dequeue doesn't have to iterate past it while ignoring it. Then we'll just start on the next
-            // command.
-            if (command.httpsRequest && !command.httpsRequest->response) {
-                httpsCommands.push_back(move(command));
-                continue;
-            }
+                // If we've dequeued a command with an incomplete HTTPS request, we move it to httpsCommands so that every
+                // subsequent dequeue doesn't have to iterate past it while ignoring it. Then we'll just start on the next
+                // command.
+                if (command.httpsRequest && !command.httpsRequest->response) {
+                    // We can't finish this transaction right now, so we roll it back. We'll restart it later when the
+                    // httpsRequest is complete.
+                    core.rollback();
 
-            // And now we'll decide how to handle it.
-            if (nodeState == SQLiteNode::MASTERING) {
-                // Now that we've peeked without finishing the command, we grab the commit mutex exclusively, so that none
-                // of our worker threads can attempt to write to the database until we're finished.
-                server._syncThreadCommitMutex.lock();
-                if (core.processCommand(command)) {
+                    // Done with the lock.
+                    server._syncThreadCommitMutex.unlock();
+
+                    // Set this aside and move on to the next command.
+                    httpsCommands.push_back(move(command));
+                    continue;
+                }
+
+                // if alreadyHadHttpsRequest is set, then we peeked this request at some point in the past and ended
+                // that transaction, so we need to start a new one, so we pass that as the `beginTransaction` flag.
+                if (core.processCommand(command, alreadyHadHttpsRequest)) {
```

Comment:
So, we may have peeked a command on a worker thread that created a httpsRequest. In this case there's no active transaction (because it was rolled back when the command was queued for the sync thread).

If not, we'll peek the command now, which will open a transaction.

So when we get to the point of calling `processCommand`, we may or may not have a transaction open.

Having `beginTransaction` and `beginConcurrentTransaction` just be no-ops if a transaction  is already in progress might be simpler. We'd just need to remove the assertions in those functions that check to make sure that there isn't a current transaction already. I don't feel like this costs us a lot of safety.